packages/boot/src/__tests__/acceptance/crud-rest-builder.acceptance.ts

```diff
@@ -0,0 +1,168 @@
+      'controllers.ProductController',
+    );
+  });
+
+  it('uses bound repository class if it exists', async () => {
+    await sandbox.copyFile(
+      resolve(__dirname, '../fixtures/product.model.js'),
+      'models/product.model.js',
+    );
+
+    await sandbox.writeTextFile(
+      'model-endpoints/product.rest-config.js',
+      `
+const {Product} = require('../models/product.model');
+module.exports = {
+  model: Product,
+  pattern: 'CrudRest',
+  dataSource: 'db',
+  basePath: '/products',
+};
+      `,
+    );
+
+    const ProductRepository = defineCrudRepositoryClass(Product);
+    app.repository(ProductRepository);
+
+    const binding = app.getBinding('repositories.ProductRepository');
+    expect(binding.key).to.eql('repositories.ProductRepository');
+
+    // Boot & start the application
+    await app.boot();
+    await app.start();
+
+    expect(app.getBinding('repositories.ProductRepository')).to.eql(binding);
+
+    expect(app.getBinding('controllers.ProductController').key).to.eql(
+      'controllers.ProductController',
+    );
+  });
+
+  it('throws if there is no base path in the config', async () => {
+    await sandbox.copyFile(
+      resolve(__dirname, '../fixtures/product.model.js'),
+      'models/product.model.js',
+    );
+
+    await sandbox.writeTextFile(
+      'model-endpoints/product.rest-config.js',
+      `
+const {Product} = require('../models/product.model');
+module.exports = {
+  model: Product,
+  pattern: 'CrudRest',
+  dataSource: 'db',
+};
+      `,
+    );
+
+    // Boot the application
+    await expect(app.boot()).to.be.rejectedWith(
+      /Missing required field "basePath" in configuration for model Product./,
+    );
+  });
+
+  it('throws if a Model is used instead of an Entity', async () => {
+    await sandbox.copyFile(
+      resolve(__dirname, '../fixtures/no-entity.model.js'),
+      'models/no-entity.model.js',
+    );
+
+    await sandbox.writeTextFile(
+      'model-endpoints/no-entity.rest-config.js',
+      `
+const {NoEntity} = require('../models/no-entity.model');
+module.exports = {
+  model: NoEntity,
+  pattern: 'CrudRest',
+  dataSource: 'db',
+  basePath: '/no-entities',
+};
+      `,
+    );
+
+    // Boot the application
+    await expect(app.boot()).to.be.rejectedWith(
+      /CrudRestController requires an Entity, Models are not supported/,
+    );
+  });
+
+  class BooterApp extends BootMixin(RepositoryMixin(RestApplication)) {
+    constructor(options?: ApplicationConfig) {
+      super(options);
+      this.projectRoot = sandbox.path;
+      this.booters(ModelApiBooter);
+      this.component(CrudRestComponent);
+    }
+  }
+
+  async function givenAppWithDataSource() {
+    app = new BooterApp({
+      rest: givenHttpServerConfig(),
+    });
+    app.dataSource(new juggler.DataSource({connector: 'memory'}), 'db');
+  }
+
+  async function stopApp() {
+    try {
+      await app.stop();
+    } catch (err) {
+      // application is booting
```

Comment:
Why do we want to hide the error? We should let `mocha` to capture and report it.

Reply:
Some of the tests are meant to throw an error during `app.boot()` so the application never starts and so if we let `mocha` capture it, it'll always throw errors every time the tests are run.

Reply:
Does it mean that `app.stop()` throws an error if the app was not started? I find that surprising, because the implementation of `stop` does seem to short-circuit to a no-op when not running. 

https://github.com/strongloop/loopback-next/blob/4192217263f5f4fc1dfe5c0a0a4deac6bcf9aad1/packages/core/src/application.ts#L285-L294

I guess fixing this problem is out of scope of this pull request? 

Have you considered checking if the app is running and don't call stop if it is not? Something along the following lines:

```ts
async function stopApp() {
  if (app.state !== 'started') return;
  await app.stop();
}
```

Maybe the `boot` method is adding more states (e.g. `booting`, `booted`) and we need to modify the checks in `start` and `stop` to account for them? IDK.

If we decide to keep ignoring the error:

Silently swallowing all errors is almost always a bad idea, because if an unexpected error happens and the rest of the test suite starts to behave differently as a result, then we have no way how to identify the problem. As a general rule:
- Try to narrow down the kinds of ignored errors as much as possible. Ideally, the error thrown by `app.stop()` should have a unique `code` property we can check.
- Print the ignored error via a `debug` log, so that when we need to troubleshoot this part, we have a way how to see all details about the ignored error.

Reply:
> Does it mean that app.stop() throws an error if the app was not started? I find that surprising, because the implementation of stop does seem to short-circuit to a no-op when not running.

If the application hasn't started, it'll throw an error from this function:

https://github.com/strongloop/loopback-next/blob/crud-rest/packages/core/src/application.ts#L225-L230

> ```ts
> async function stopApp() {
>   if (app.state !== 'started') return;
>   await app.stop();
> }
> ```

I think this is a good way to check (since `app.state` is otherwise `booting` and in that case we don't want to stop the application), so I'll use it.

Reply:
> I think this is a good way to check (since app.state is otherwise booting and in that case we don't want to stop the application), so I'll use it.

Sounds good :+1: